src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+                        // Or AccessToken is not null, mean token based authentication is used.
+                        if ((context.ConnectionContext.ConnectionString != null
+                            && context.ConnectionContext.ConnectionString.Authentication != SqlAuthenticationMethod.NotSpecified)
+                            || context.ConnectionContext.AccessTokenInBytes != null || context.ConnectionContext.AccessTokenCallback != null)
+                        {
+                            context.ConnectionContext.FedAuthRequired = payload[payloadOffset] == 0x01;
+                        }
+                        break;
+
+                    default:
+                        Debug.Fail("UNKNOWN option in ConsumePreLoginHandshake, option:" + option);
+                        // DO NOTHING FOR THESE UNKNOWN OPTIONS
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES;
+                        break;
+                }
+
+                if (offset < payload.Length)
+                {
+                    option = payload[offset++];
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructs the Prelogin packet and sends it to the server.
+        /// </summary>
+        /// <param name="context">The Prelogin handler context</param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private static async Task CreatePreLoginAndSend(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(context.ConnectionContext.TdsStream != null, "A Tds Stream is expected");
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+            tdsStream.PacketHeaderType = TdsStreamPacketType.PreLogin;
+
+            // Initialize option offset into payload buffer
+            // 5 bytes for each option (1 byte length, 2 byte offset, 2 byte payload length)
+            int offset = (int)PreLoginOptions.NUMOPT * 5 + 1;
+
+            byte[] payload = new byte[(int)PreLoginOptions.NUMOPT * 5 + TdsEnums.MAX_PRELOGIN_PAYLOAD_LENGTH];
+            int payloadLength = 0;
+
+            byte[] instanceName = new byte[1];
+
+            for (int option = (int)PreLoginOptions.VERSION; option < (int)PreLoginOptions.NUMOPT; option++)
+            {
+                int optionDataSize = 0;
+
+                // Fill in the option
+                await tdsStream.WriteByteAsync((byte)option, isAsync, ct);
+
+                // Fill in the offset of the option data
+                await tdsStream.WriteByteAsync((byte)((offset & 0xff00) >> 8), isAsync, ct); // send upper order byte
+                await tdsStream.WriteByteAsync((byte)(offset & 0x00ff), isAsync, ct); // send lower order byte
```

Comment:
Rather than doing this, just write to the correct indexes of the payload. It'll eliminate any overhead from the awaits, and means you only need to perform one write at the end.